pennylane/debugging.py

```diff
@@ -105,3 +108,75 @@ def get_snapshots(*args, **kwargs):
         return dbg.snapshots
 
     return get_snapshots
+
+
+class PLDB(pdb.Pdb):
+    """Custom debugging class integrated with Pdb."""
+
+    __active_dev = []
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the debugger, and set custom prompt string."""
+        super().__init__(*args, **kwargs)
+        self.prompt = "[pldb]: "
+
+    @classmethod
+    def valid_context(cls):
+        """Determine if the debugger is called in a valid context.
+
+        Raises:
+            TypeError: Can't call breakpoint outside of a qnode execution
+            TypeError: Device not supported with breakpoint
+        """
+
+        if not qml.queuing.QueuingManager.recording() or not cls.is_active_dev():
+            raise TypeError("Can't call breakpoint outside of a qnode execution")
+
+        if cls.get_active_device().name not in ("default.qubit", "lightning.qubit"):
+            raise TypeError("Device not supported with breakpoint")
+
+    @classmethod
+    def add_device(cls, dev):
+        """Add a device to the global active device list.
+
+        Args:
+            dev (Union[Device, "qml.devices.Device"]): The active device
+        """
+        cls.__active_dev.append(dev)
+
+    @classmethod
+    def get_active_device(cls):
+        """Return the active device.
+
+        Raises:
+            ValueError: No active device to get
+
+        Returns:
+            Union[Device, "qml.devices.Device"]: The active device
+        """
+        if not cls.is_active_dev():
+            raise ValueError("No active device to get")
+
+        return cls.__active_dev[0]
+
+    @classmethod
+    def is_active_dev(cls):
+        """Determine if there is currently an active device.
+
+        Returns:
+            bool: True if there is an active device
+        """
+        return bool(cls.__active_dev)
+
+    @classmethod
+    def reset_active_dev(cls):
+        """Reset the global active device list (to empty)."""
+        cls.__active_dev = []
+
+
+def breakpoint():
+    """Launch the custom PennyLane debugger."""
```

Comment:
What is the plan with docs here - normally PRs should have detailed docs, especially for major user-facing functions like this. Though if we have a story to track a subsequent docs revisit and polish, I'm fine with that too.

Reply:
There is a story to revisit the docs and polish. I will add a little more detail to the breakpoint docs in this PR, but once all of the functionality (measuring states, queuing gates in debug mode, etc.) are added in the subsequent PRs, then we can finalize its docstring.

Reply:
Minor point that we can discuss in a docs PR:

We don't have a quickstart (.rst) for `debugging` yet. And it doesn't seem like we make quickstarts for any of the high level files in the `pennylane` folder. Maybe there's a question here about whether debugging functionality should go in `pennylane/debugging.py` or in ``pennylane/debugging/debugging.py``